File: week4/python/OddEvenLinkedList_328.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def oddEvenList(self, head: ListNode) -> ListNode:
        if not head:
            return head
        
        odd_node = head
        root_even_node = head.next
        even_node = root_even_node

        while even_node and even_node.next:
            odd_node.next = even_node.next
            odd_node = odd_node.next
            even_node.next =  odd_node.next
            even_node = even_node.next
            
        odd_node.next = root_even_node

        return head
        

File: week4/java/OddEvenLinkedList_328.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode oddEvenList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        
        ListNode odd_node = head;
        ListNode even_node = head.next;
        ListNode root_even_node = even_node;
        
        while (even_node != null && even_node.next != null) {
            odd_node.next = even_node.next;
            odd_node = odd_node.next;
            even_node.next = odd_node.next;
            even_node = even_node.next;
        }
        odd_node.next = root_even_node;
        return head;
    }
}
